src/core/mac/sub_mac.cpp

```diff
@@ -276,6 +276,52 @@ void SubMac::CslSample(void)
 }
 #endif // OPENTHREAD_CONFIG_MAC_CSL_RECEIVER_ENABLE
 
+#if OPENTHREAD_CONFIG_MAC_CSL_DEBUG_ENABLE
+void SubMac::LogReceived(RxFrame *aFrame)
+{
+    Address  dst;
+    char     logString[64];
+    int32_t  drift;
+    uint32_t sampleTime, ahead, after;
+
+    IgnoreError(aFrame->GetDstAddr(dst));
+
+    VerifyOrExit((dst.GetType() == Address::kTypeShort && dst.GetShort() == GetShortAddress()) ||
+                 (dst.GetType() == Address::kTypeExtended && dst.GetExtended() == GetExtAddress()));
+
+    LogWarn("Received frame in state (SubMac %s, CSL %s), timestamp %lu", StateToString(mState),
+            mIsCslSampling ? "CslSample" : "CslSleep",
+            ToUlong(static_cast<uint32_t>(aFrame->mInfo.mRxInfo.mTimestamp)));
+
+    VerifyOrExit(mState == kStateCslSample);
+
+    GetCslWindowEdges(ahead, after);
+    ahead -= kCslReceiveTimeAhead;
+
+    sampleTime = RadioSupportsReceiveTiming() ? mCslSampleTime.GetValue() - mCslPeriod * kUsPerTenSymbols
```

Comment:
I tested with the simulator platform (without receive-timing support) and the sampleTime was always roughly one CSL period too high. (So the reported drift was always a large negative value).

Using `mCslSampleTime.GetValue() - mCslPeriod * kUsPerTenSymbols` was already better (small, positive drift values). However I would expect a drift of 0 reported in the simulation case mostly so there's still something off - either in the code or in my CSL timing validation script, or both.

Reply:
> small, positive drift values

Which number, approximately?

Reply:
Yes, actually the subtraction of the period time should happen in all cases.

With regards to the logged drift, we should take into account that even with perfect transmitter timing, there will be an error due to the CSL phase rounding from microseconds to units of 160 us. Could this be the case?